pkg/kubelet/kubelet.go

```diff
@@ -2972,22 +3070,47 @@ func (kl *Kubelet) CheckpointContainer(
 		return fmt.Errorf("container %v not found", containerName)
 	}
 
-	options.Location = filepath.Join(
+	checkpointDirectory := filepath.Join(
 		kl.getCheckpointsDir(),
+		string(podUID),
+	)
+
+	checkPointDirectoryInfo, err := os.Lstat(checkpointDirectory)
+	if err == nil && !checkPointDirectoryInfo.Mode().IsDir() {
+		return fmt.Errorf("checkpoint destination %s needs to be a directory", checkpointDirectory)
+	} else if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error stating checkpoint destination %s:%w", checkpointDirectory, err)
+	} else if err != nil && os.IsNotExist(err) {
+		if err = os.Mkdir(checkpointDirectory, 0700); err != nil {
+			return fmt.Errorf("error creating checkpoint directory %s:%w", checkpointDirectory, err)
+		}
+	}
+
+	kl.maxCheckpointsPerContainerLock.Lock()
+	options.Location = filepath.Join(
+		checkpointDirectory,
 		fmt.Sprintf(
-			"checkpoint-%s-%s-%s.tar",
+			"checkpoint-%s-%s-%d-%0*d.tar",
 			podFullName,
 			containerName,
-			time.Now().Format(time.RFC3339),
```

Comment:
> I was thinking for the forensic case

Even if the story around checkpoint/restore is the forensic case, this does not mean this is what people use it for. There are many use cases and the forensic use case is just one.

> the caller would be responsible for the clean up if need (probably with a listener) and the use would be rare.

This is a strange argument from my point of view, that the user is responsible for the clean up. At this point it feels the complete PR is questioned and this argument comes very late in the lifetime of this PR. This PR has been reworked completely multiple times since the first posting 7 months ago. This is a feature people have been asking for because they see it as a problem and do not want to clean up manually. This is one of the first things people are asking for during conference talks.

Not sure what the goal of this comment is, sorry.

Reply:
The reason for limiting to the forensics case was probably to avoid slow walking the managed checkpoint use cases. Until there was a WG or whatnot put in place to design how kubernetes would manage the checkpoints for the other cases for what people would use it for.  

At this point the user requested the checkpoint, vs kubelet creating the checkpoint based on a pod policy / contract. Asking kubelet to garbage collect that which it did not ask to be created implies kubelet knows why the containers are being checkpoints for these pods.

If the rule is to keep 10 which 10, the last ten the first 10 what if the last 10 are all created during failure modes? If this is for rolling backups.. you may want one per month one per week...going back to the las month to drop the per week from that month etc..

Reply:
.. today was the first day I heard about this PR... Apologies for the frustration.  I agree 100%, overuse of the checkpointing end point without having a management design is a problem. I didn't expect over use to be a problem for the sig-node approved forensics use case.  That's all I mean.  For logs we've used similar designs to this one, in kubelet, to employ rolling log models for long lived containers.  So I'm trying to understand this rolling checkpoint idea.. to map it to a use case. If someone is wanting to do rolling backups ok.. I would get the use case, but even in that case I would want to have a discussion about how we do it.  Checkpoint 1 + delta + delta for example would be orders of magnitude better from a resource consumption perspective.